src/xserver_zmq_split.cpp

```diff
@@ -33,7 +40,32 @@ namespace xeus
         , p_publisher(new xpublisher(context,
                                      std::bind(&xserver_zmq_split::serialize_iopub, this, std::placeholders::_1),
                                      config.m_transport, config.m_ip, config.m_iopub_port))
-        , p_shell(new xshell(context, config.m_transport, config.m_ip ,config.m_shell_port, config.m_stdin_port, this))
+        , p_shell(new xshell_default(context, config.m_transport, config.m_ip, config.m_shell_port, config.m_stdin_port, this))
+        , m_control_thread()
+        , m_hb_thread()
+        , m_iopub_thread()
+        , m_shell_thread()
+        , m_error_handler(eh)
+        , m_control_stopped(false)
+    {
+        p_controller->connect_messenger();
+    }
+
+    // In the case where an event loop is provided
+    xserver_zmq_split::xserver_zmq_split(zmq::context_t& context,
+                                         const xconfiguration& config,
+                                         nl::json::error_handler_t eh,
+                                         std::shared_ptr<uvw::loop> loop_ptr,
+                                         std::unique_ptr<xhook_base> hook)
+        : p_auth(make_xauthentication(config.m_signature_scheme, config.m_key))
+        , p_controller(new xcontrol(context, config.m_transport, config.m_ip ,config.m_control_port, this))
+        , p_heartbeat(new xheartbeat(context, config.m_transport, config.m_ip, config.m_hb_port))
+        , p_publisher(new xpublisher(context,
+                                     std::bind(&xserver_zmq_split::serialize_iopub, this, std::placeholders::_1),
+                                     config.m_transport, config.m_ip, config.m_iopub_port))
+        , p_shell(new xshell_uv(context, config.m_transport, config.m_ip,
+                                config.m_shell_port, config.m_stdin_port, this,
+                                loop_ptr, std::move(hook)))
```

Comment:
This should be factorized with a private delegating constructor. The public constructors should initialize the right shell object, and forward it (and everything else) to the delegating constructor.

Reply:
I updated this, but I'm not 100% if this is what you meant. Please let me know. Thanks!